src/model/model.ts

```diff
@@ -1677,6 +1694,145 @@ class InputModel {
+        this.aiChatTextAreaRef = textAreaRef;
+        this.aiChatWindowRef = chatWindowRef;
+    }
+    
+    setAIChatFocus() {
+        if (this.aiChatTextAreaRef != null && this.aiChatTextAreaRef.current != null) {
+            this.aiChatTextAreaRef.current.focus();
+        }
+    }
+    
+    grabCodeSelectSelection() {
+        if(this.codeSelectSelectedIndex >= 0 && this.codeSelectSelectedIndex < this.codeSelectBlockRefArray.length) {
+            let curBlockRef = this.codeSelectBlockRefArray[this.codeSelectSelectedIndex]
+            let codeText = curBlockRef.current.innerText;
+            codeText = codeText.replace(/\n$/, "") // remove trailing newline        
+            let newLineValue = this.getCurLine() + " " + codeText;
+            this.setCurLine(newLineValue);
+            this.giveFocus();
+        }
+    }
+     
+    addCodeBlockToCodeSelect(blockRef: React.RefObject<HTMLElement>): number {
+        let rtn = -1;
+        rtn = this.codeSelectBlockRefArray.length;            
+        this.codeSelectBlockRefArray.push(blockRef);
+        this.codeSelectDeselectAll();
+        return rtn;
+    }
+    
+    codeSelectDeselectCodeBlock(blockIndex: number) {
+        if(blockIndex >= 0 && blockIndex < this.codeSelectBlockRefArray.length) {
+            let currentRef = this.codeSelectBlockRefArray[blockIndex].current;
+            if(currentRef != null) {
+                currentRef.classList.remove("selected");
+            }
+        }  
+    }
+    
+
+    setCodeSelectSelectedCodeBlock(blockIndex:number) { 
+        if(blockIndex >= 0 && blockIndex < this.codeSelectBlockRefArray.length) {
+            this.codeSelectDeselectCodeBlock(this.codeSelectSelectedIndex);
+            this.codeSelectSelectedIndex = blockIndex;
+            let currentRef = this.codeSelectBlockRefArray[blockIndex].current;
+            if(currentRef != null) {
+                currentRef.classList.add("selected");
+                if(this.aiChatWindowRef != null && this.aiChatWindowRef.current != null) {
+                    let chatWindowTop = this.aiChatWindowRef.current.scrollTop;
+                    let chatWindowBottom = chatWindowTop + this.aiChatWindowRef.current.clientHeight;
+                    let elemTop = currentRef.offsetTop;
+                    let elemBottom = elemTop - currentRef.offsetHeight;
+                    let elementIsInView = (elemBottom < chatWindowBottom && elemTop > chatWindowTop);
+                    console.log(chatWindowBottom, chatWindowTop, elemBottom, elemTop);
+                    if(!elementIsInView) {
+                        this.aiChatWindowRef.current.scrollTop = elemBottom - (this.aiChatWindowRef.current.clientHeight / 3);                        
+                    }
+                }
+            }
+            this.setAIChatFocus();
+        }  
+    }
+    
+    codeSelectDecrementCodeBlock() {
```

Comment:
This part is a bit wierd. I add code blocks to the end of the data structure so that each one preserves an index value, but when the user is cycling through them, we want to look through them in backwards order of time added, so from array length -1 to index 0. so "increment code select" selects the code block one above, but it actually selects the code block stored in the data structure at current selected index - 1. I don't wanna create tech debt and have one of you guys have to suffer through trying to change this in a few months lol, so I should look into adding comments or changing this to be more readable

Reply:
addressed, I renamed the functions to make them a little more clear, and added a comment